### PR TITLE
API Remove custom logic in favour of Form::saveInto()

### PR DESCRIPTION
**Pull request by GuySartorelli**
I don't think this warrants a note in the changelog beyond what will be automatically generated for the removed API - though feel free to disagree and if so I'll add a docs PR.

## Issue
- https://github.com/silverstripe/silverstripe-elemental/issues/1268

### CODE REVIEW

File: src/Controllers/ElementalAreaController.php [hunk 1]
```diff
@@ -354,25 +353,19 @@ public function save(array $data, Form $form): HTTPResponse
 
     /**
      * Remove the pseudo namespaces that were added to form fields by the form factory
-     *
-     * @param array $data
-     * @param int $elementID
-     * @return array
      */
-    public static function removeNamespacesFromFields(array $data, $elementID)
+    private function removeNamespacesFromFields(Form $form, int $elementID): void
```
Comment (GuySartorelli) [inline src/Controllers/ElementalAreaController.php:357]: This method wasn't used anywhere else, so no need for it to be public nor static.

File: src/Controllers/ElementalAreaController.php [hunk 2]
```diff
@@ -298,8 +298,7 @@ public function getElementForm(): Form
     }
 
     /**
-     * Arrive here from FormRequestHandler::httpSubmission() during a POST request to
-     * /admin/linkfield/linkForm/<LinkID>
+     * Arrive here from FormRequestHandler::httpSubmission() during a POST request.
```
Comment (GuySartorelli) [inline src/Controllers/ElementalAreaController.php:301]: Unrelated change here to remove copy/paste error that references linkfield

File: src/Forms/EditFormFactory.php [hunk 3]
```diff
@@ -103,4 +100,22 @@ protected function namespaceFields(FieldList $fields, array $context)
             $field->setName($namespacedName);
         }
     }
+
+    /**
+     * Remove the pseudo namespaces that were added in namespaceFields()
+     */
+    public function removeNamespaceFromFields(FieldList $fields, array $context): void
```
Comment (GuySartorelli) [inline src/Forms/EditFormFactory.php:107]: Using the same method signature as the corresponding `namespaceFields()` since they're essentially the reverse of one another.

Comment (emteknetnz) [inline src/Forms/EditFormFactory.php:107]: Add a unit test

File: src/Controllers/ElementalAreaController.php [hunk 4]
```diff
@@ -331,11 +330,15 @@ public function save(array $data, Form $form): HTTPResponse
             $this->jsonError(403);
         }
 
-        // Remove the namespace prefixes that were added by EditFormFactory
-        $dataWithoutNamespaces = static::removeNamespacesFromFields($data, $element->ID);
+        // Remove the namespace prefixes
+        $factory = Injector::inst()->get(EditFormFactory::class);
+        $factory->removeNamespaceFromFields($form->Fields(), ['Record' => $element]);
```
Comment (emteknetnz) [inline src/Controllers/ElementalAreaController.php:335]: The way these methods are being called on a factory class (whose purpose to a to create objects) looks weird as they're essentially utility methods.  Change them to be statics and don't instantiate the factory, or just move them both to a standalone utility class as statics

**Review by emteknetnz — CHANGES_REQUESTED** [review]

File: src/Controllers/ElementalAreaController.php (hunk 4 above)
Comment (GuySartorelli) [reply src/Controllers/ElementalAreaController.php:335]: I don't want to do that, because the factory is injected and the existing `namespaceFields()` method could be altered in a project-specific customisation. Breaking that customisation option for the sake of "it looks weird" doesn't sit well with me.

The namespacing is intrinsic to the building of this specific form, so it makes sense for the factory to own that process.

**Comment by GuySartorelli** [comment]
None of the things you've pointed out in this review block the CMS 5 PRs. Please merge the CMS 5 PRs.

File: src/Forms/EditFormFactory.php (hunk 3 above)
Comment (GuySartorelli) [reply src/Forms/EditFormFactory.php:107]: There's no test for `namespaceFields()` nor for the old `removeNamespacesFromFields()` method on `ElementalAreaController`... these methods are both indirectly tested already.
But to avoid ping pong, I'll add tests for both of these methods after rebasing on top of the CMS 5 PR.

File: src/Controllers/ElementalAreaController.php (hunk 4 above)
Comment (emteknetnz) [reply src/Controllers/ElementalAreaController.php:335]: Looking at this again, renaming the fields on an existing form twice in order to save it seems kind of wrong

Wouldn't it make more sense to just get the factory to create an entirely new form without the namespaces, load the submitted $data into that and then call $newFormWithoutNamespaces->saveInto($element)

That would allow us to keep keep namespaceFields() as a protected method and we wouldn't even need removeNamespacesFromFields()

**Update: ignore this, just occured to me the the submitted data will include namespaces**

**Review by emteknetnz — CHANGES_REQUESTED** [review]

File: src/Forms/EditFormFactory.php (hunk 3 above)
Comment (GuySartorelli) [reply src/Forms/EditFormFactory.php:107]: Done

File: tests/Forms/EditFormFactoryTest.php [hunk 5]
```diff
@@ -4,15 +4,18 @@
...
+use SilverStripe\Forms\LiteralField;
+use SilverStripe\Forms\TextField;
 
 class EditFormFactoryTest extends SapphireTest
 {
     protected static $fixture_file = 'EditFormFactoryTest.yml';
 
     public function testFormFieldsHaveNamespaces()
     {
-        /** @var ElementContent $record */
```
Comment (GuySartorelli) [inline tests/Forms/EditFormFactoryTest.php:15]: Unrelated, but that's not needed anymore due to the addition of generics

File: tests/Forms/EditFormFactoryTest.php [hunk 6]
```diff
@@ -21,4 +24,42 @@ public function testFormFieldsHaveNamespaces()
...
         $this->assertNotNull($fields->dataFieldByName('PageElements_' . $record->ID . '_Title'));
     }
+
+    public function testNamespaceFields(): void
+    {
+        $record = $this->objFromFixture(ElementContent::class, 'content_block');
+        $factory = new EditFormFactory();
+        $fields = new FieldList([
+            new TextField('FieldOne'),
+            new LiteralField('IgnoredField', ''),
```
Comment (GuySartorelli) [inline tests/Forms/EditFormFactoryTest.php:34]: `LiteralField` is ignored because it's not a "data field" so it doesn't affect form submission

File: tests/Forms/EditFormFactoryTest.php [hunk 7]
```diff
@@ -21,4 +24,42 @@ public function testFormFieldsHaveNamespaces()
...
+        $this->assertSame($expectedNames, $fields->column('Name'));
+    }
+
+    public function testRemoveNamespaceFromFields(): void
+    {
+        $record = $this->objFromFixture(ElementContent::class, 'content_block');
+        $factory = new EditFormFactory();
+        $fields = new FieldList([
+            new TextField('PageElements_' . $record->ID . '_FieldOne'),
+            new LiteralField('IgnoredField', ''),
```
Comment (GuySartorelli) [inline tests/Forms/EditFormFactoryTest.php:53]: This is how the literal field will look when passed in, so we're just including it here to validate it doesn't get mangled

**Review by emteknetnz — APPROVED** [review]
